### DISCUSSION TITLE
Spack in Airgapped Environments

### DISCUSSION THREAD

**Discussion by teaguesterling** [spack/spack#44573] [Q&A]
I'm curious about what approaches people are taking to use spack effectivity in airgapped environments. I've had decent success by creating a local mirror with simpler packages. However, I've run into lots of headaches with a couple of different kinds of packages and was curious what patterns people are using.

The sorts of packages that are problematic:
 * Those that checkout directly from github. The problems here fall into two categories: when the build system tries to inspect git metadata, which isn't mirrored (e.g. version identification); and those that need submodules, which seem to have mixed success in mirroring. In these cases I often create a custom tarball and add it as a new custom version, but this is a pain.
 * Those where versions have conflicts on specific hardware/os configurations as it can be difficult to concreting the appropriate dependencies if the machine you are building the mirror on is significantly different from the architecture of the build host. In these cases, I wrestle with specific lock files or the concretizer to manually mirror the versions I need.
 * Those where the build system actually tries to download dependencies itself. In this case, I'll try to patch the and add a spack resource and coax the build system into believing it's already done the download. This is pretty tedious, error-prone, nor reusable.

Is this a lost cause, is something fundamental I'm missing, or is the a roadmap that I could help contribute to in order to implement some of these needs?

For the record, py-torch is all of the above :-).

**Comment by tgamblin** [comment]
Creating a mirror is the way to go, and then there is an extra step you need to do to put the bootstrap binaries in your mirror:
* https://spack.readthedocs.io/en/latest/bootstrapping.html#creating-a-mirror-for-air-gapped-systems

For the three issues above:
1. Right now git checkouts are mirrored as tarballs, but this isn't great for a number of reasons.  We can't re-checksum the tarballs on the other side, and as you say we throw out the `.git` directory. Also, if you ask for several commits from the same repo, they're stored several times.  We should probably introduce logic that bundles up the latest version of the repo with full history, and ensures that any checked out commits are in it. Also submodules should probably get some kind of special handling but I am not sure what that should look like.  @psakievich is working on features so that *every* git ref (e.g. potentially moving targets like branches and tags) are resolved to commits. A rework of git mirroring leveraging that would be welcome. 
2. You *can* do spack mirror create `--all` to get all versions of the things you're mirroring, so that you could potentially re-concretize on the other side. You could also choose a more generic architecture (e.g. x86_64_v3) to build for and sync binaries instead of tarballs.  I'd have to dig into specifics to say much more about this one.
3. We typically patch the build for this - if `py-torch` is downloading deps, we should really ensure that it doesn't happen.. I don't know another great way to ensure that it can be built reliably on the others side.  I don't think it's a lost cause but yes it can require some package maintenance. @adamjstewart might be interested to know if `py-torch` is downloading extra stuff.

**Reply by teaguesterling** [reply]
Thanks for the detailed reply! I apologize in advance for the even more wordy response. 

I'm currently creating mirrors and patching as needed (which works great 95% of the time --- even if a bit of fiddling is required for the git-based packages. The extreme cases often get messy and start to look more like dark arts than controlled builds. If there's a need for help on this, I'm interested in improving this for usability, reproducibility, and to make lives inside the airgap better.

For the third item, there are a few categories of packages I've come across, each that may have varying levels of difficulty to address:

A) Packages that run curl/wget to some discrete set of additional resources or patches on their own for convenience and make the installation easier for users. These are usually pretty easy to handle. As long as you can figure out what is being downloaded where, you can add a resource and patch out the download. I ran into this while packaging HAIL and I believe RDKit will attempt this under certain conditions. 

B) Packages that are written in languages that have first-class package managers that retrieve external resources on l-demand as part of the build process. Both Rust and Go do this, making building those packages from mirrors difficult. I ran into this with the SingularityCE package and some rust packages I can't recall right now. NodeJS/NPM too, but that's not as much of an issue I suppose. 

C) Packages that run some sort of explicit large recursive dependency installation through a package manager or script, such as calling pip as part of the install. I believe this was what I was running into in PyTorch. In some ways, this is similar to (B), but it seems different to me as it's not a corr feature of the language (one could argue).

For (A), it's pretty straightforward as part of packaging to handle. 

For (B), I suspect it could be handled by ensuring all of the specific dependencies are available in Spack. It's hard to figure out what they all are given the "quietness" of it working outside an airgap, but the tooling is standard and could probably be audited statically. We may be able to have some sort of a Cargo.toml checker for rust, for example.

For (C), these are often the hardest since they're just occurring (possibly inconsistently) as part of the build process. It would be nice to know when this is happening, but it'd be even nicer to have a way to capture what's requested and either mirror, review, or white list it. 

I was considering creating an extension that would turn on a slim HTTP(S) proxy that would be configured so that (at least well-behaved) tool chains would pasd through it. It would be more of an admin tool than anything else. It could be activated within spack to monitor installations and see if they retrieve external resources. It could also be used to create a list of the URLs and checksums to retain some of Spack's integrity checking. Lastly (and most interesting to me), it likely could also cache everything downloaded during an installation and create a resource bundle that could be re-served through the HTTP(S) proxy within the airgapped environment. That would at least provide a path forward for mirroring those extra-tricky packages.

**Comment by psakievich** [comment]
@teaguesterling lots of good points here. It would be nice if we could require packages that do (C) to place these fetchers in a member function so it can be overridden from an inherited package.  One of the main ways I've seen projects handle these sorts of air-gap constraints is to either a) write their own package.py in a custom repo, or b) create a package in a custom repo that inherits from the builtin repo.  The former is typically for teams who have their usage of a package dialed or run into issues with the builtin that can't be overridden via inheritance.

This clearly won't handle all the issues, and I like your idea of a monitoring tool.  It would be nice if we could add this to CI and audit/score packages based on their compatibility with air-gapped systems.  I suspect many package maintainers don't think about these issues at all, but would if they were made aware at the PR level.
